src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+	rangeTableList);
+static void GetAllUniqueIndexes(Form_pg_index indexForm, List **uniqueIndexes);
+static RTEToSubqueryConverterReference * GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+																	   RTEToSubqueryConverterContext
+																	   *
+																	   rteToSubqueryConverterContext,
+																	   PlannerRestrictionContext
+																	   *
+																	   plannerRestrictionContext,
+																	   Oid
+																	   resultRelationId);
+static void GetRangeTableEntriesFromJoinTree(Node *joinNode, List *rangeTableList,
+											 List **joinRangeTableEntries);
+static void RemoveFromRTEToSubqueryConverterContext(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext, Oid relationId);
+static bool FillLocalAndDistributedRTECandidates(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+	RTEToSubqueryConverterReference **
+	localRTECandidate,
+	RTEToSubqueryConverterReference **
+	distributedRTECandidate);
+
+/*
+ * ConvertUnplannableTableJoinsToSubqueries gets a query and the planner
+ * restrictions. As long as the query is not plannable by router planner,
+ * it converts either a local or distributed table to a subquery.
+ */
+void
+ConvertUnplannableTableJoinsToSubqueries(Query *query,
+										 RecursivePlanningContext *context)
+{
+	List *rangeTableList = NIL;
+	GetRangeTableEntriesFromJoinTree((Node *) query->jointree, query->rtable,
+									 &rangeTableList);
+	Oid resultRelationId = GetResultRelationId(query);
+	if (!ShouldConvertLocalTableJoinsToSubqueries(rangeTableList, resultRelationId))
+	{
+		return;
+	}
+
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext =
+		CreateRTEToSubqueryConverterContext(
+			context, rangeTableList);
+
+	RTEToSubqueryConverterReference *rteToSubqueryConverterReference =
+		GetNextRTEToConvertToSubquery(query->jointree, rteToSubqueryConverterContext,
+									  context->plannerRestrictionContext,
+									  resultRelationId);
+
+	PlannerRestrictionContext *plannerRestrictionContext =
+		FilterPlannerRestrictionForQuery(context->plannerRestrictionContext, query);
+	while (rteToSubqueryConverterReference && !IsRouterPlannable(query,
+																 plannerRestrictionContext))
+	{
+		ReplaceRTERelationWithRteSubquery(
+			rteToSubqueryConverterReference->rangeTableEntry,
+			rteToSubqueryConverterReference->restrictionList,
+			rteToSubqueryConverterReference->
+			requiredAttributeNumbers,
+			context);
+		RemoveFromRTEToSubqueryConverterContext(rteToSubqueryConverterContext,
+												rteToSubqueryConverterReference->
+												rangeTableEntry->relid);
+		rteToSubqueryConverterReference =
+			GetNextRTEToConvertToSubquery(query->jointree, rteToSubqueryConverterContext,
+										  context->plannerRestrictionContext,
+										  resultRelationId);
+	}
+}
+
+
+/*
+ * GetResultRelationId gets the result relation id from query
+ * if it exists.
+ */
+static Oid
+GetResultRelationId(Query *query)
+{
```

Comment:
Minor: We  seem  to have `ModifyQueryResultRelationId`, which does the  same?

Reply:
it errors if it is not a modify result relation, we don't want to error

Reply:
Optional note: To me, it is slightly confusing to call `GetResultRelationId ` for a non-modification query anyway. So, it'd be easier to read if we use this function with `IsModifyCommand(originalQuery) && ..`